js/src/common/states/PaginatedListState.ts

```diff
@@ -0,0 +1,196 @@
+  number: number;
+  items: TModel[];
+}
+
+export interface PaginationLocation {
+  page: number;
+  startIndex?: number;
+  endIndex?: number;
+}
+
+export default abstract class PaginatedListState<T extends Model> {
+  protected location!: PaginationLocation;
+  protected offset: number;
+
+  protected pages: Page<T>[] = [];
+  protected params: any = {};
+
+  protected initialLoading: boolean = false;
+  protected loadingPrev: boolean = false;
+  protected loadingNext: boolean = false;
+
+  hasPrev: boolean = false;
+  hasNext: boolean = false;
+
+  protected constructor(offset: number = 20) {
+    this.offset = offset;
+  }
+
+  abstract get type(): string;
+
+  protected parseResults(pageNum, results: T[]) {
+    const page = {
+      number: pageNum,
+      items: results,
+    };
+
+    if (this.isEmpty() || pageNum > this.getNextPageNumber() - 1) {
+      this.pages.push(page);
+    } else {
+      this.pages.unshift(page);
+    }
+
+    this.hasNext = !!results.payload?.links?.next;
+    this.hasPrev = !!results.payload?.links?.prev;
+
+    m.redraw();
+  }
+
+  /**
+   * Load a new page of results.
+   */
+  protected loadPage(page = 1): Promise<T[]> {
+    const params = this.requestParams();
+    params.page = { offset: this.offset * (page - 1) };
+
+    if (Array.isArray(params.include)) {
+      params.include = params.include.join(',');
+    }
+
+    return app.store.find(this.type, params);
+  }
+
+  public hasItems(): boolean {
+    return !!this.getAllItems().length;
+  }
+
+  public isEmpty(): boolean {
+    return !this.isInitialLoading() && !this.hasItems();
+  }
+
+  public getLocation(): PaginationLocation {
+    return this.location;
+  }
+
+  protected getAllItems(): T[] {
+    return this.pages.map((pg) => pg.items).flat();
+  }
+
+  public isInitialLoading(): boolean {
+    return this.initialLoading;
+  }
+  public isLoadingPrev(): boolean {
+    return this.loadingPrev;
+  }
+  public isLoadingNext(): boolean {
+    return this.loadingNext;
+  }
+
+  /**
+   * Get the parameters that should be passed in the API request.
+   * Do not include page offset unless subclass overrides loadPage.
+   *
+   * @abstract
+   * @see loadPage
+   */
+  protected requestParams(): any {
+    return {};
+  }
+
+  /**
+   * Stored state parameters.
+   */
+  public getParams(): any {
+    return this.params;
+  }
+
+  /**
+   * Update the `this.params` object, calling `refresh` if they have changed.
+   * Use `requestParams` for converting `this.params` into API parameters
+   *
+   * @param newParams
+   * @see requestParams
+   */
+  public refreshParams(newParams) {
+    if (this.isEmpty() || Object.keys(newParams).some((key) => this.requestParams()[key] !== newParams[key])) {
+      this.params = newParams;
+
+      return this.refresh();
+    }
+  }
+
+  public refresh(page: number = 1) {
+    this.initialLoading = true;
+    this.loadingPrev = false;
+    this.loadingNext = false;
+
+    m.redraw();
```

Comment:
Shouldn't we clear it here?

Reply:
I'd say no, in case there's an error when fetching results?

Reply:
We only do this because of our logic with whether the previous page matches `IndexPage`, I think... to force the refresh when calling `refreshParams`.

Perhaps we should just return a loading spinner if `isInitialLoading()` is true, that'd avoid any issues I feel like...

Reply:
We need to keep it, as without it, clicking the refresh button shows no UI confirnation.